### ISSUE TITLE
Need a performant way to compose within foreach

### ISSUE THREAD

**Issue by AdamWillden** [knockout/knockout#1638]
I'm running into performance issues using nested [knockout foreach](http://knockoutjs.com/documentation/foreach-binding.html) bindings that heavily utilise [knockout components](http://knockoutjs.com/documentation/component-overview.html). In my case the bindings produce a dynamic table (columns and rows) with optionally modifiable cell contents.

I can use [knockout-repeat](https://github.com/mbest/knockout-repeat) to marginally speed things up but the fact it works on the outer html is causing me issues with components I've made, such that I have to do things to my code that 'breaks' the nice encapsulation I've done with components. The reason I'm forced to break encapsulation is because I'm using table (tbody, tr, td) elements within which you can't put anything 'unusual' (like a div).

Below you can see a stripped down version of what I'm doing currently. I've taken some styling bindings and component parameters out. It produces a nest as follows:
`*` **`<tbody>` foreach (50-100 items)**
`* *` **virtual component `<tr>`**
`* * *` **foreach (<15 items)**
`* * * *` **virtual component `<td>`**

Below is some of the table markup

``` html
<thead>
    <tr data-bind="foreach: visibleColumns">
        <!-- column header markup-->
    </tr>
</thead>
<tbody data-bind="template: { foreach: pageData }">
    <!-- ko component: {
        name: 'table-row',
        params: {
             sourceData: $data,
            columns: $parent.visibleColumns,
            selectMethod: $parent.config.selectMethod,
            isSelected: $parent.getSelectionIndicator($data)
            /*...etc...*/
        }
    } --><tr><td></td></tr><!-- /ko -->
</tbody>
```

and 'table-row' component markup

``` html
<tr>
<!-- ko if: selectMethod === 'check' -->
    <td class="omsui-table-rowSelect" data-bind="toggle: { data: isSelected, stylize: false }">
    {{#if: isSelected}}
        <i class="fa fa-fw fa-lg fa-check-square-o"></i>
    {{/if}}
    {{#ifnot: isSelected}}
        <i class="fa fa-fw fa-lg fa-square-o"></i>
    {{/if}}
    </td>
<!-- /ko -->
<!-- ko foreach: columns -->
    <!-- ko component: {
        name: 'table-cell',
        params: {
            sourceData: $parent.sourceData,
            column: $data,
            paths: $parent.paths,
            pathStatuses: $parent.pathStatuses
        }
    } --><td></td><!-- /ko -->
<!-- /ko -->
</tr>
```

Because the component _and_ foreach bindings both call applyBindings I end up with double the apply binding calls than I actually need. I ~~am~~ was looking for something that operates like the foreach binding (on the inner HTML) but behaves like the repeat binding (doesn't apply bindings to repeated nodes).

I now question whether components were expected to be used so intensively, they allow me to break down my code so tremendously that I wouldn't want to do without but I understand it's going to bring in some performance overheads.

**tl;dr & finally** 

I've come to wonder with the advent components whether there may be a need for a binding in the form of `foreach : { data: pageData, compose: {...} }`

**Comment by brianmhunt** [comment]
@AdamWillden  (or anyone else following this) — Since you have tested the timing out, could you give something a shot for me, please?

Would you mind changing the stock/master branch as follows:

Add to the beginning of [loaderRegistry.js](https://github.com/AdamWillden/knockout/blob/master/src/components/loaderRegistry.js):

```javascript
var animationRequest = window.requestAnimationFrame
    || window.webkitRequestAnimationFrame
    || window.mozRequestAnimationFrame
    || window.msRequestAnimationFrame
    || function(cb) { return window.setTimeout(cb, 1000 / 60); };
```

Then replace [line 17](https://github.com/AdamWillden/knockout/blob/master/src/components/loaderRegistry.js#L17):

```javascript
setTimeout(function() { callback(cachedDefinition.definition); }, 0);		
```

with

```javascript
animationRequest(function() { callback(cachedDefinition.definition); })
```

and testing out the performance of how that compares to the current as well as the batched versions.

I take the lead of the above from the [FastDom](https://github.com/wilsonpage/fastdom/blob/master/index.js#L16) project, and animation frames are [surprisingly well supported](https://developer.mozilla.org/en/docs/Web/API/window.requestAnimationFrame). I am under the belief that the `requestAnimationFrame` may do the batching for us, and execute reflows coincidentally with other queued DOM changes. But who knows! I think it would just be interesting to see how it compares, and it ought to be a very simple test (if you already have the code to run the comparison).

Thanks & Cheers.

**Comment by brigand** [comment]
setTimeout 0 is almost always a poor choice.  Either you want 'async but as fast as possible', e.g. setImmediate, process.nextTick, or a browser trick like postMessage/onmessage or microtasks via mutation events (which should be abstracted into a ko util, with fallbacks) – or you want an animation tick like brianmhunt suggested.

The difference, is an animation tick could take 0.01667s, or could run in 3 hours when the user switches back to your tab.  

Or you can race a requestAnimationFrame and setTimeout 100, if you want 'it'll run soonish, ideally on an animation tick'.

**Comment by AdamWillden** [comment]
@brianmhunt I'm trying it now

**Comment by AdamWillden** [comment]
I am changing from showing 5 items to 263 items on 8 columns. I also ensure all 263 items have been shown once already to ensure the required components are pre-loaded.

```
16.2s stock
10.5s batch
12.1s anim
```

I definitely need to profile my own app code again as it seems to have gotten bad while developing, I know someone is probably exclaiming seeing those times for 263 items. Regardless, the differences between the stock, batch and animation are still quite clear.

**Comment by SteveSanderson** [comment]
We now have @mbest's task scheduler in place and resolving component load requests, which eliminates the `setTimeout(..., 0)`. Hopefully this will lead to some improvements with better batching.

@AdamWillden, is there any chance you could let us know whether the current code in `master` behaves better for you?

**Comment by idrake** [comment]
@SteveSanderson In my experience V3.3.0 solved the my layout jank in my grids where every cell is a component.  I'm very happy with the fix.

**Comment by AdamWillden** [comment]
@SteveSanderson I'll check ASAP

**Comment by AdamWillden** [comment]
Happy to say that code in #1738 has helped and is just as performant as what was being attempted in #1659. Also I'm using @brianmhunt's fastForEach plugin with the noContext option to remove the additional contexts. 

I'd still like to keep this issue open or open another to get a step further with cleaner markup (see first/second comment) and a more optimized foreach component binding performance wise. If you believe this is an edge case requirement then please close, the current solutions do work well.
